test/onnx/exporter/test_building.py

```diff
@@ -0,0 +1,147 @@
+# Owner(s): ["module: onnx"]
+"""Unit tests for the _building module."""
+
+from __future__ import annotations
+
+import numpy as np
+import onnxscript
+from onnxscript import ir
+
+import torch
+from torch.onnx._internal.exporter import _building, _tensors
+from torch.testing._internal import common_utils
+
+
+class TestOpRecorder(common_utils.TestCase):
+    def setUp(self):
+        self.opset_version = 17
+        self.opset = onnxscript.values.Opset("", self.opset_version)
+        self.recorder = _building.OpRecorder(opset=self.opset, constant_farm={})
+
+        self.model = ir.Model(
+            graph=ir.Graph(
+                [],
+                [],
+                nodes=[],
+                opset_imports={
+                    "": self.opset_version,
+                },
+                name="main_graph",
+            ),
+            ir_version=9,
+            producer_name="pytorch",
+            producer_version=torch.__version__,
+        )
+
+    def test_skippable_castlike(self):
+        input_x = _tensors.SymbolicTensor(opset=self.opset, name="input_x")
+        input_x.dtype = ir.DataType.FLOAT
+
+        input_y = _tensors.SymbolicTensor(opset=self.opset, name="input_y")
+        input_y.dtype = ir.DataType.FLOAT
+
+        with onnxscript.evaluator.default_as(
+            tracer := self.recorder,
+        ):
+            cast = self.opset.CastLike(input_y, input_x)
+            _ = self.opset.Add(input_x, cast)
+
+        self.assertEqual(len(tracer.nodes), 1)
+        self.assertEqual(tracer.nodes[0].op_type, "Add")
+
+    def test_castlike_to_cast(self):
+        input_x = _tensors.SymbolicTensor(opset=self.opset, name="input_x")
+        input_x.dtype = ir.DataType.FLOAT
+
+        input_y = _tensors.SymbolicTensor(opset=self.opset, name="input_y")
+        input_y.dtype = ir.DataType.INT64
+
+        with onnxscript.evaluator.default_as(
+            tracer := self.recorder,
+        ):
+            cast = self.opset.CastLike(input_y, input_x)
+            _ = self.opset.Add(input_x, cast)
+
+        self.assertEqual(len(tracer.nodes), 2)
+        self.assertEqual(tracer.nodes[0].op_type, "Cast")
+        self.assertEqual(tracer.nodes[1].op_type, "Add")
+
+    def test_process_python_constant(self):
+        input_x = _tensors.SymbolicTensor(
+            opset=self.opset, name="input_x", shape=ir.Shape([2, 3, 4])
+        )
+        new_shape = [3, 2, 4]
+
+        with onnxscript.evaluator.default_as(
+            tracer := self.recorder,
+        ):
+            _ = self.opset.Reshape(input_x, new_shape)
```

Comment:
Should we test different types of constants?

Reply:
I think for tracer tests, we only needs to take care of the tracer.nodes. Do you have a specific test case in mind? I guess you are referring to test on `_determine_input_dtype`, but I think the function is pretty straight forward though.